File: frontend/src/types/product.ts
import { productFileUrl } from "@/lib/api/config";
import type { SiteCurrency } from "@/types/siteConfig";

export type Product = {
  title: string;
  slug: string;
  reviews?: number;
  /** Price in minor units (e.g. cents). */
  price: number;
  /** Discounted price in minor units (e.g. cents). */
  discountedPrice: number;
  id: number;
  description?: string;
  currency?: string;
  minorUnit?: number;
  imgs?: {
    thumbnails: string[];
    previews: string[];
  };
};

/** Backend product_file / variant file item */
export type ApiProductFile = {
  file_id: string;
  product_id: number;
  variant_id: number | null;
  content_type: string;
  position: number;
  is_primary: boolean;
};

/** Backend API product shape (GET /products) */
export type ApiProduct = {
  productId: number;
  sku: string;
  name: string;
  description: string;
  slug: string;
  price: number;
  isActive: boolean;
  inStock: boolean;
  productFile?: ApiProductFile | null;
};

export function mapApiProductToProduct(
  api: ApiProduct,
  currency?: SiteCurrency | null
): Product {
  const url = productFileUrl(api.productFile?.file_id);
  return {
    id: api.productId,
    title: api.name,
    slug: api.slug,
    description: api.description,
    price: api.price,
    discountedPrice: api.price,
    currency: currency?.code,
    minorUnit: currency?.minorUnit,
    reviews: 0,
    imgs: url
      ? {
          thumbnails: [url],
          previews: [url],
        }
      : undefined,
  };
}

File: backend/internal/model/product.go
package model

import (
	"time"
)

type Products []Product

type Product struct {
	ID          uint         `db:"product_id" json:"productId"`
	MerchantID  *uint64      `db:"merchant_id" json:"merchantId"`
	SKU         string       `db:"sku" json:"sku"`
	Name        string       `db:"name" json:"name"`
	Description *string      `db:"description" json:"description"`
	Slug        string       `db:"slug" json:"slug"`
	PriceAmount uint         `db:"price_amount" json:"price"`
	IsActive    bool         `db:"is_active" json:"isActive"`
	InStock     bool         `db:"in_stock" json:"inStock"`
	CreatedAt   time.Time    `db:"created_at" json:"-"`
	UpdatedAt   time.Time    `db:"updated_at" json:"-"`
	DeletedAt   *time.Time   `db:"deleted_at" json:"-"`
	ProductFile *ProductFile `json:"productFile,omitempty"`
	Categories  Categories   `json:"category,omitempty"`
}
